xsnap/sources/xsnap-worker.c

```diff
@@ -541,7 +584,10 @@ int main(int argc, char* argv[])
 					c_exit(E_IO_ERROR);
 				}
 				if (snapshot.error == 0) {
-					int writeError = fxWriteOkay(toParent, meterIndex, machine, "", 0);
+					// Allows us to format up to 999,999,999,999 bytes (1TiB - 1)
+					char fsize[13];
+					int fsizeLength = snprintf(fsize, sizeof(fsize) - 1, "%d", stream.size);
```

Comment:
Why `sizeof(fsize) - 1`?  `snprintf` won't NUL-terminate the buffer if it hits this limit, so you have 1 byte of initialized stack in it.  In that case, you might actually mean:

```suggestion
					int fsizeLength = snprintf(fsize, sizeof(fsize) - 1, "%d", stream.size);
					fsize[fsizeLength] = '\0';
```

Or on the other hand, if you don't need fsize to be NUL-terminated, maybe you mean:

```suggestion
					int fsizeLength = snprintf(fsize, sizeof(fsize), "%d", stream.size);
```

Reply:
I'm guessing this was copied from the `snprintf()` in `fxWriteOkay()`, around line 813, which is probably mine. My read of the `snprintf` man page (on my Mac):

> The snprintf() and vsnprintf() functions will write at most size-1 of the characters printed into the output string (the size'th character then gets the terminating ‘\0’); if the return value is greater than or equal to the size argument, the string was too short and some of the printed characters were discarded.  The output is always null-terminated, unless size is 0.

is that using `sizeof(fsize)-1` means `fsize` will always be NUL-terminated, even if the argument is too big (the string will be truncated in that case, but it will still be NUL-terminated). I remember worrying about clobbering the stack, and wanted to make sure that even if I miscalculated how large the buffer needed to be, the worst case would be corrupted meter data and not crashing or corrupting the execution.

Reply:
Another man quote:
> The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.

So I assume that yes this is being overly cautious. It does also mean it will only write 11 chars right now, aka 100GiB not 1TiB...

I'll fix both.

Reply:
Ah, looks like C99 defines the always-terminate behaviour.  I wrote C where not all platforms were C99, so had to be more defensive against the various footguns.